### DISCUSSION TITLE
Error accessing Reality Model in ContextShare

### DISCUSSION THREAD

**Discussion by ovaisa** [iTwin/itwinjs-core#2091] [Q&A] [answered]
I am trying to move my code to the latest (2.19.6) iTwin (from 2.15.2).  I created a new application as my existing application didn't seem to work with 2.19.6 (not sure if this statement is correct but I was having problems so I created a new application).  

I am getting http error when I try to query reality data info from ContextShare.  I have  realitydata:modify and realitydata:read specified in my scope and also enabled in application.

I recall that last time that Bentley had to do something manually (add realitydata scopes) to the application to get it to work (wondering if this is same issue).

![Scopes](https://user-images.githubusercontent.com/83007711/129928329-3fc23f35-4a2c-4b17-ba8a-3993bb3a40a2.JPG)

The following queryRealityData() call is failing

    const queryCriteria: RealityDataQueryCriteria = { contextId: this.iModel.contextId! };
    const availableModels: ContextRealityModelProps[] = await queryRealityData(queryCriteria);

=======
GET https://api.bentley.com/contextshare/v1/Repositories/S3MXECPlugin--d20678d8-ad8c-4732-a32f-5ec62e2befc0/S3MX/RealityData?project=d20678d8-ad8c-4732-a32f-5ec62e2befc0&$filter=Type+eq+%27RealityMesh3DTiles%27+or+Type+eq+%27OPC%27+or+Type+eq+%27Terrain3DTiles%27+or+Type+eq+%27OMR%27+or+Type+eq+%27Cesium3DTiles%27 401
push../node_modules/superagent/lib/client.js.Request._end @ client.js:890
push../node_modules/superagent/lib/client.js.Request.end @ client.js:744
(anonymous) @ request-base.js:266
push../node_modules/superagent/lib/request-base.js.RequestBase.then @ request-base.js:248

=======
The network request associated with this call 

Request URL: https://api.bentley.com/contextshare/v1/Repositories/S3MXECPlugin--d20678d8-ad8c-4732-a32f-5ec62e2befc0/S3MX/RealityData?project=d20678d8-ad8c-4732-a32f-5ec62e2befc0&$filter=Type+eq+%27RealityMesh3DTiles%27+or+Type+eq+%27OPC%27+or+Type+eq+%27Terrain3DTiles%27+or+Type+eq+%27OMR%27+or+Type+eq+%27Cesium3DTiles%27
Request Method: GET
Status Code: 401 
Remote Address: 13.76.135.165:443
Referrer Policy: strict-origin-when-cross-origin
access-control-allow-origin: *
access-control-expose-headers: Request-Context,Strict-Transport-Security,Mas-Server,X-Correlation-ID,Mas-Request-Id,Content-Length,Set-Cookie,WWW-Authenticate,Date
cache-control: must-revalidate, no-cache, private
content-length: 0
date: Wed, 18 Aug 2021 15:13:42 GMT
mas-request-id: 4c50c471-b2b1-49bd-b6d2-8f062145edbe
mas-server: Bentley-WSG/4.0.13,Bentley-WebAPI/2.9
pragma: no-cache
request-context: appId=cid-v1:58d31e1b-ef83-49c6-8bb1-930b36fdbc40
set-cookie: ARRAffinity=20648448b6fdcca8782b22c056e67800fe46cf86507329d514d3d0935c0caf13;Path=/;HttpOnly;Secure;Domain=connect-realitydataservices.bentley.com
set-cookie: ARRAffinitySameSite=20648448b6fdcca8782b22c056e67800fe46cf86507329d514d3d0935c0caf13;Path=/;HttpOnly;SameSite=None;Secure;Domain=connect-realitydataservices.bentley.com
strict-transport-security: max-age=63072000
www-authenticate: SAML realm="Bentley WSApi", Token realm="Bentley WSApi", Bearer realm="Bentley WSApi"
x-correlation-id: 4c50c471-b2b1-49bd-b6d2-8f062145edbe
:authority: api.bentley.com
:method: GET
:path: /contextshare/v1/Repositories/S3MXECPlugin--d20678d8-ad8c-4732-a32f-5ec62e2befc0/S3MX/RealityData?project=d20678d8-ad8c-4732-a32f-5ec62e2befc0&$filter=Type+eq+%27RealityMesh3DTiles%27+or+Type+eq+%27OPC%27+or+Type+eq+%27Terrain3DTiles%27+or+Type+eq+%27OMR%27+or+Type+eq+%27Cesium3DTiles%27
:scheme: https
accept: application/json
accept-encoding: gzip, deflate, br
accept-language: en-US,en;q=0.9
authorization:  ... REMOVED ....
origin: http://localhost:3000
referer: http://localhost:3000/
sec-ch-ua: "Chromium";v="92", " Not A;Brand";v="99", "Google Chrome";v="92"
sec-ch-ua-mobile: ?0
sec-fetch-dest: empty
sec-fetch-mode: cors
sec-fetch-site: cross-site
user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.159 Safari/537.36
x-correlation-id: 4c50c471-b2b1-49bd-b6d2-8f062145edbe
project: d20678d8-ad8c-4732-a32f-5ec62e2befc0
$filter: Type eq 'RealityMesh3DTiles' or Type eq 'OPC' or Type eq 'Terrain3DTiles' or Type eq 'OMR' or Type eq 'Cesium3DTiles'
﻿
TreeHandler.tsx:43 getNodesCount:No Parent-count: 7
TreeHandler.tsx:62 getNodes for: Area

=======
The decoded token is:

HEADER:ALGORITHM & TOKEN TYPE

{
  "alg": "RS256",
  "kid": "BentleyIMS",
  "pi.atm": "a8me"
}

PAYLOAD:DATA

bf": 1629297635,
  "usage_country_iso": "US",
  "auth_time": 1629297935,
  "name": "ovais.ashraf@trivor.com",
  "family_name": "Ashraf",
  "email": "ovais.ashraf@trivor.com",
  "exp": 1629301538
}
{
  "scope": [
    "email",
    "openid",
    "profile",
    "organization",
    "itwinjs",
    "realitydata:modify",
    "realitydata:read",
    "forms:read",
    "issues:read",
    "issues:modify"
  ],
  "client_id": "spa-noKUboL1uvtUEdQ4TQKIG9b62",
  "aud": [
    "https://ims.bentley.com/as/token.oauth2",
    "https://imsoidc.bentley.com/as/token.oauth2",
    "https://imsoidc.bentley.com/resources",
    "itwinjs-services",
    "bentley-api-management"
  ],
  "sub": "64c796be-0f75-4cf2-98d5-1144bc069242",
  "role": "SELECT_DOWNLOAD",
  "subject": "64c796be-0f75-4cf2-98d5-1144bc069242",
  "iss": "https://ims.bentley.com",
  "preferred_username": "ovais.ashraf@trivor.com",
  "given_name": "Ovais",
  "sid": "wkrwTlZg9tfMqpnYG6iPsat7JAQ.SU1TLUJlbnRsZXktU0c.HR0G.Mv6mhpXMGAD3ab8qeG9asn5sn",
  "nbf": 1629297635,
  "usage_country_iso": "US",
  "auth_time": 1629297935,
  "name": "ovais.ashraf@trivor.com",
  "family_name": "Ashraf",
  "email": "ovais.ashraf@trivor.com",
  "exp": 162930
}

**Comment by EricPaquetB** [comment, marked as answer]
The one thing that I think is missing is that you are not using the new IMS Authority url when getting your token. 

You can look here for an example of how to configure it, https://github.com/imodeljs/imodeljs/discussions/1699#discussioncomment-1139495.

I was also told that a blog post to help users migrate should be published soon...

**Reply by ovaisa** [reply]
Hi Eric,

I was not trying to migrate authentication of my existing app (as per #1699) , instead I created a new iTwin app for this so I expect it should be (and seems to be - pls see decoded token below) using the new authentication based on IMS authority.  Note the "aud" and "iss" fields, both seem to point to IMS.  Reality data related scopes are included and are set in the client app too.

For demonstration of the issue, I created a new app from iTwin react template today and added code to query for reality data.  Please see attached xviewer.zip at bottom of this post.  My changes are:  I added 'onIModelConnected' parameter to the 'Viewer' and added its callback (all changes in App.tsx).  The callback code is:  

  const onIModelReady = async (imodel: IModelConnection) => {
    if (!imodel) return;
    IModelApp.viewManager.onViewOpen.addOnce(async (vp: ScreenViewport) => {
      const availableModels = await queryRealityData({ contextId: imodel.contextId! });  // ***THIS CALL RETURNS 401 ERROR***
    });
  }
 
Note that I am using the Bayview Plant model with this app.  The project doesn't have any reality data associated with it, but the call should not fail (and doesn't fail in my old project).  It should zero reality data models, rather than 401 error.  If for debugging purposes, anyone needs access to my test project, I'll be happy to add them in, please let me know.

=== Error ====

![Error](https://user-images.githubusercontent.com/83007711/130055923-aff05398-9806-43ab-954d-1eb7f7beaf58.JPG)


==== Decoded token =====
{
  "alg": "RS256",
  "kid": "BentleyIMS",
  "pi.atm": "a8me"
}.{
  "scope": [
    "email",
    "openid",
    "profile",
    "organization",
    "itwinjs",
    "forms:modify",
    "forms:read",
    "issues:read",
    "issues:modify",
    "realitydata:modify",
    "realitydata:read",
    "library:modify",
    "library:read",
    "imodels:read",
    "imodels:modify",
    "storage:modify",
    "storage:read",
    "users:read",
    "projects:modify",
    "projects:read"
  ],
  "client_id": "spa-noKUboL1uvtUEdQ4TQKIG9b62",
  "aud": [
    "https://ims.bentley.com/as/token.oauth2",
    "https://imsoidc.bentley.com/as/token.oauth2",
    "https://imsoidc.bentley.com/resources",
    "itwinjs-services",
    "bentley-api-management"
  ],
  "sub": "64c796be-0f75-4cf2-98d5-1144bc069242",
  "role": "SELECT_DOWNLOAD",
  "subject": "64c796be-0f75-4cf2-98d5-1144bc069242",
  "iss": "https://ims.bentley.com",
  "preferred_username": "ovais.ashraf@trivor.com",
  "given_name": "Ovais",
  "sid": "lZMm4KVHQDj8KR0fiJufavGthgE.SU1TLUJlbnRsZXktU0c.Hiqy.vKlNxSw78tXENd8ke0Memdnew",
  "nbf": 1629366671,
  "usage_country_iso": "US",
  "auth_time": 1629366971,
  "name": "ovais.ashraf@trivor.com",
  "family_name": "Ashraf",
  "email": "ovais.ashraf@trivor.com",
  "exp": 1629370574
} 

==== client scopes ====

![Scopes](https://user-images.githubusercontent.com/83007711/130050238-9ea40d87-c2b0-437c-8de5-38a3d05b3f38.JPG)

==== XVIEWER.ZIP ====

[xviewer.zip](https://github.com/imodeljs/imodeljs/files/7013733/xviewer.zip)

**Comment by EricPaquetB** [comment]
Ovais, we'll need to investigate this issue further. One possibility is that the realitydata scope is not correctly delegated...

Will get back to you.

-Eric

**Reply by ovaisa** [reply]
Hi Eric,

Can I request an update on this.  Reality Models are central to our workflow and because of this issue, our progress is stranded.

Thank you

**Comment by EricPaquetB** [comment]
Ovais,
Sorry about the long wait. I contacted another team to investigate this issue, because I did not find anything on my side… I’ll push to get feedback.

Thanks,
Eric

**Reply by EricPaquetB** [reply]
Ovais,
This issue requires a fix on our side. We'll try to deploy the fix as soon as possible. In the meantime, I check if some workaround could unblock you (but it's not likely, unfortunately).

Regards,
Eric

**Reply by ovaisa** [reply]
Thanks for the update.  Yes, if you can find a workaround, it would be really appreciated.

I had similar problem with accessing reality data a while back and Bentley added some scope privileges (or something) manually behind the scenes in the clientapp which at least helped me continue with my development work.  If there is such a work around, it will be very helpful.

Thanks, Ovais

**Comment by EricPaquetB** [comment]
Sure, I remember. The issue is still related to scopes, but the situation is a bit different this time. In any case, we’re actively working on pushing a fix.

Eric

**Reply by EricPaquetB** [reply]
Ovais, we have a fix for your issue. We'll push it in the coming days. There does not seem to be a workaround in the meantime, unfortunately.
Sorry for the delay,
Eric

**Reply by ovaisa** [reply]
Thank you Eric for following up, glad to hear that fix will be available in the next push.  If it is possible to share the timeline of that push, it will help me manage our user expectation.

**Comment by ovaisa** [comment]
Hi Eric,

I wanted to point out from last time we ran into this issue.  If you recall, the access to reality model access was working in development mode (localhost:3000) but it was not working when I deployed the site on Netlify.  I recall that Glen commented that the access worked within Bentley domain/network but it was an issue when accessing from outside Bentley domain (that is why it was not caught in testing by Bentley).  Not sure if that was really the issue, I hope that this is checked before pushing the new version (as it would cause more delays on our side if their were similar problems accessing from outside Bentley domain) 

Thanks,

Ovais

**Reply by EricPaquetB** [reply]
Hi Ovais,
The issue that you mention was really caused by how we handled scopes. And we now have a similar problem. Actually, a change was committed to the API layer, but I don't know if this made it to Prod. Could you try to re-run your app (no need to update) and see if there is some improvement on your side ?

Thanks,
Eric

**Comment by ovaisa** [comment]
I am afraid I am still getting the same error

![Error](https://user-images.githubusercontent.com/83007711/131352254-439afa34-20de-43fe-bf7c-55a6cc7a6ecb.JPG)


Request URL: https://api.bentley.com/contextshare/v1/Repositories/S3MXECPlugin--d20678d8-ad8c-4732-a32f-5ec62e2befc0/S3MX/RealityData?project=d20678d8-ad8c-4732-a32f-5ec62e2befc0&$filter=Type+eq+%27RealityMesh3DTiles%27+or+Type+eq+%27OPC%27+or+Type+eq+%27Terrain3DTiles%27+or+Type+eq+%27OMR%27+or+Type+eq+%27Cesium3DTiles%27
Request Method: GET
Status Code: 401 
Remote Address: 13.76.135.165:443
Referrer Policy: strict-origin-when-cross-origin
access-control-allow-origin: *
access-control-expose-headers: Request-Context,Strict-Transport-Security,Mas-Server,X-Correlation-ID,Mas-Request-Id,Content-Length,Set-Cookie,WWW-Authenticate,Date
cache-control: must-revalidate, no-cache, private
content-length: 0
date: Mon, 30 Aug 2021 13:58:49 GMT
mas-request-id: 2d908417-c951-4aab-a09f-5b22c2a626c3
mas-server: Bentley-WSG/4.0.13,Bentley-WebAPI/2.9
pragma: no-cache
request-context: appId=cid-v1:58d31e1b-ef83-49c6-8bb1-930b36fdbc40
set-cookie: ARRAffinity=20648448b6fdcca8782b22c056e67800fe46cf86507329d514d3d0935c0caf13;Path=/;HttpOnly;Secure;Domain=connect-realitydataservices.bentley.com
set-cookie: ARRAffinitySameSite=20648448b6fdcca8782b22c056e67800fe46cf86507329d514d3d0935c0caf13;Path=/;HttpOnly;SameSite=None;Secure;Domain=connect-realitydataservices.bentley.com
strict-transport-security: max-age=63072000
www-authenticate: SAML realm="Bentley WSApi", Token realm="Bentley WSApi", Bearer realm="Bentley WSApi"
x-correlation-id: 2d908417-c951-4aab-a09f-5b22c2a626c3
:authority: api.bentley.com
:method: GET
:path: /contextshare/v1/Repositories/S3MXECPlugin--d20678d8-ad8c-4732-a32f-5ec62e2befc0/S3MX/RealityData?project=d20678d8-ad8c-4732-a32f-5ec62e2befc0&$filter=Type+eq+%27RealityMesh3DTiles%27+or+Type+eq+%27OPC%27+or+Type+eq+%27Terrain3DTiles%27+or+Type+eq+%27OMR%27+or+Type+eq+%27Cesium3DTiles%27
:scheme: https
accept: application/json
accept-encoding: gzip, deflate, br
accept-language: en-US,en;q=0.9
authorization:  ***DELETED***

origin: http://localhost:3000
referer: http://localhost:3000/
sec-ch-ua: "Chromium";v="92", " Not A;Brand";v="99", "Google Chrome";v="92"
sec-ch-ua-mobile: ?0
sec-fetch-dest: empty
sec-fetch-mode: cors
sec-fetch-site: cross-site
user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.159 Safari/537.36
x-correlation-id: 2d908417-c951-4aab-a09f-5b22c2a626c3
project: d20678d8-ad8c-4732-a32f-5ec62e2befc0
$filter: Type eq 'RealityMesh3DTiles' or Type eq 'OPC' or Type eq 'Terrain3DTiles' or Type eq 'OMR' or Type eq 'Cesium3DTiles'

**** DECODED TOKEN ****
{
  "alg": "RS256",
  "kid": "BentleyIMS",
  "pi.atm": "a8me"
}.{
  "scope": [
    "email",
    "openid",
    "profile",
    "organization",
    "itwinjs",
    "forms:modify",
    "forms:read",
    "issues:read",
    "issues:modify",
    "realitydata:modify",
    "realitydata:read",
    "library:modify",
    "library:read",
    "imodels:read",
    "imodels:modify",
    "storage:modify",
    "storage:read",
    "users:read",
    "projects:modify",
    "projects:read"
  ],
  "client_id": "spa-noKUboL1uvtUEdQ4TQKIG9b62",
  "aud": [
    "https://ims.bentley.com/as/token.oauth2",
    "https://imsoidc.bentley.com/as/token.oauth2",
    "https://imsoidc.bentley.com/resources",
    "itwinjs-services",
    "bentley-api-management"
  ],
  "sub": "64c796be-0f75-4cf2-98d5-1144bc069242",
  "role": "SELECT_DOWNLOAD",
  "subject": "64c796be-0f75-4cf2-98d5-1144bc069242",
  "iss": "https://ims.bentley.com",
  "preferred_username": "ovais.ashraf@trivor.com",
  "given_name": "Ovais",
  "sid": "sjG_eCaArM1hXfLdOrmoem55htg.SU1TLUJlbnRsZXktU0c.LOQB.fUMAxkZmNs2URbUyo8bF0DKOa",
  "nbf": 1630331614,
  "usage_country_iso": "US",
  "auth_time": 1630331914,
  "name": "ovais.ashraf@trivor.com",
  "family_name": "Ashraf",
  "email": "ovais.ashraf@trivor.com",
  "exp": 1630335517
}

**Reply by EricPaquetB** [reply]
Ovais,
Maybe the issue is related to client registration. Could you try to create a new application (so that we're sure that it's registered with the new client), and see if you can access reality data ? And make sure that the itwinjs scope is present.

**Reply by ovaisa** [reply]
Created a new client, same error

Request URL: https://api.bentley.com/contextshare/v1/Repositories/S3MXECPlugin--d20678d8-ad8c-4732-a32f-5ec62e2befc0/S3MX/RealityData?project=d20678d8-ad8c-4732-a32f-5ec62e2befc0&$filter=Type+eq+%27RealityMesh3DTiles%27+or+Type+eq+%27OPC%27+or+Type+eq+%27Terrain3DTiles%27+or+Type+eq+%27OMR%27+or+Type+eq+%27Cesium3DTiles%27
Request Method: GET
Status Code: 401 
Remote Address: 13.76.135.165:443
Referrer Policy: strict-origin-when-cross-origin
access-control-allow-origin: *
access-control-expose-headers: Request-Context,Strict-Transport-Security,Mas-Server,X-Correlation-ID,Mas-Request-Id,Content-Length,Set-Cookie,WWW-Authenticate,Date
cache-control: must-revalidate, no-cache, private
content-length: 0
date: Mon, 30 Aug 2021 18:27:22 GMT
mas-request-id: 6ad70f8f-0834-4c69-9a00-14e1c1e81c89
mas-server: Bentley-WSG/4.0.13,Bentley-WebAPI/2.9
pragma: no-cache
request-context: appId=cid-v1:58d31e1b-ef83-49c6-8bb1-930b36fdbc40
set-cookie: ARRAffinity=20648448b6fdcca8782b22c056e67800fe46cf86507329d514d3d0935c0caf13;Path=/;HttpOnly;Secure;Domain=connect-realitydataservices.bentley.com
set-cookie: ARRAffinitySameSite=20648448b6fdcca8782b22c056e67800fe46cf86507329d514d3d0935c0caf13;Path=/;HttpOnly;SameSite=None;Secure;Domain=connect-realitydataservices.bentley.com
strict-transport-security: max-age=63072000
www-authenticate: SAML realm="Bentley WSApi", Token realm="Bentley WSApi", Bearer realm="Bentley WSApi"
x-correlation-id: 6ad70f8f-0834-4c69-9a00-14e1c1e81c89
:authority: api.bentley.com
:method: GET
:path: /contextshare/v1/Repositories/S3MXECPlugin--d20678d8-ad8c-4732-a32f-5ec62e2befc0/S3MX/RealityData?project=d20678d8-ad8c-4732-a32f-5ec62e2befc0&$filter=Type+eq+%27RealityMesh3DTiles%27+or+Type+eq+%27OPC%27+or+Type+eq+%27Terrain3DTiles%27+or+Type+eq+%27OMR%27+or+Type+eq+%27Cesium3DTiles%27
:scheme: https
accept: application/json
accept-encoding: gzip, deflate, br
accept-language: en-US,en;q=0.9


{
  "alg": "RS256",
  "kid": "BentleyIMS",
  "pi.atm": "a8me"
}.{
  "scope": [
    "email",
    "openid",
    "profile",
    "organization",
    "itwinjs",
    "forms:modify",
    "forms:read",
    "issues:read",
    "issues:modify",
    "realitydata:modify",
    "realitydata:read",
    "library:modify",
    "library:read",
    "imodels:read",
    "imodels:modify",
    "storage:modify",
    "storage:read",
    "users:read",
    "projects:modify",
    "projects:read"
  ],
  "client_id": "spa-nafYwXhKRXicpYZMMsGtvIYE6",
  "aud": [
    "https://ims.bentley.com/as/token.oauth2",
    "https://imsoidc.bentley.com/as/token.oauth2",
    "https://imsoidc.bentley.com/resources",
    "itwinjs-services",
    "bentley-api-management"
  ],
  "sub": "64c796be-0f75-4cf2-98d5-1144bc069242",
  "role": "SELECT_DOWNLOAD",
  "subject": "64c796be-0f75-4cf2-98d5-1144bc069242",
  "iss": "https://ims.bentley.com",
  "preferred_username": "ovais.ashraf@trivor.com",
  "given_name": "Ovais",
  "sid": "EFFYeFBO8RtYVj5clutid2Zgzos.SU1TLUJlbnRsZXktU0c.LSLw.FU5YH7uW16rztyzElRiiLg86h",
  "nbf": 1630347724,
  "usage_country_iso": "US",
  "auth_time": 1630348024,
  "name": "ovais.ashraf@trivor.com",
  "family_name": "Ashraf",
  "email": "ovais.ashraf@trivor.com",
  "exp": 1630351627
}

**Reply by EricPaquetB** [reply]
Ovais,
Just to give you an update, we're working actively on fixing this issue. We currently have some problems with our deployment pipelines, which prevent us to release, but we progress. We consider this issue as a priority.

-Eric

**Reply by EricPaquetB** [reply]
Ovais, the release of the fix is complete. Please retry to run your app and let me know how it works.

-Eric

**Reply by ovaisa** [reply]
Thank you Eric, yes it is working!!!  Much appreciate your diligence and follow up to get this resolved...

**Reply by EricPaquetB** [reply]
Great ! Very glad to hear it !

Sorry for the trouble, hope it goes more smoothly in the future.

-Eric
